src/Microsoft.AspNet.Routing/Template/TemplateBinder.cs

```diff
@@ -220,14 +220,21 @@ public string BindValues(IDictionary<string, object> acceptedValues)
                             // we won't necessarily add it to the URI we generate.
                             if (!context.Buffer(converted))
                             {
-                                return null;
+                                    return null;
                             }
                         }
                         else
                         {
                             if (!context.Accept(converted))
                             {
-                                return null;
+                                if (part.IsOptional && segment.Parts[j - 1].IsOptionalSeperator)
+                                {
+                                    context.Remove(segment.Parts[j - 1].Text);
+                                }
```

Comment:
If it's something that can't happen, add an assert `Debug.Assert(...)` and a brief comment explaining that the parser does it. This lets us know if the code breaks, but without having to really *handle* a case that can never happen.